File: main/3d-surface-area/3d-surface-area.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub read_record {
    return split /\s+/msx, <>;
}

sub read_row {
    return (0, read_record, 0);
}

my ($height, $width) = read_record;
my @zero_row = (0) x ($width + 2); # OK to use twice as the board never changes
my @board = (\@zero_row, (map { [read_row] } (1..$height)), \@zero_row);

my $area = $height * $width * 2; # top and bottom

for my $i (0..$height) {
    for my $j (0..$width) {
        $area += abs $board[$i]->[$j] - $board[$i]->[$j + 1];
        $area += abs $board[$i]->[$j] - $board[$i + 1]->[$j];
    }
}

print "$area\n";

File: main/3d-surface-area/3d-surface-area.py
#!/usr/bin/env python3

def read_record():
    return map(int, input().split())

def read_row():
    return (0, *read_record(), 0)

height, width = read_record()
nrows = height + 2
ncols = width + 2
zero_row = (0,) * ncols
board = (zero_row, *(read_row() for _ in range(height)), zero_row)

area = height * width * 2 # top and bottom

for i in range(nrows - 1):
    for j in range(ncols - 1):
        area += (abs(board[i][j] - board[i][j + 1])
               + abs(board[i][j] - board[i + 1][j]))

print(area)
